File: modules/Conversation/Models/WeChatMessage.php
<?php

declare(strict_types=1);

namespace Modules\Conversation\Models;

use Catch\Base\CatchModel as Model;
use Illuminate\Database\Eloquent\Relations\BelongsTo;

/**
 * 企业微信聊天消息模型
 * 对应 wechat_message 表
 * 
 * @property int $id 自增主键
 * @property string $msgid 企业微信消息ID
 * @property int $seq 顺序号
 * @property int $msgtime 消息时间戳（毫秒）
 * @property string $action 消息动作：send/recall
 * @property string|null $roomid 群聊ID
 * @property string|null $conversation_id 会话ID
 * @property string $conversation_type 会话类型：single/group
 * @property string $from_user 发送者userid
 * @property array|null $to_list 接收人列表
 * @property string $msgtype 消息类型
 * @property array|null $content 消息主要内容（JSON）
 * @property array|null $raw_json 原始微信返回JSON
 * @property string $create_time 记录创建时间
 */
class WeChatMessage extends Model
{
    /**
     * 表名
     *
     * @var string
     */
    protected $table = 'wechat_message';

    /**
     * 主键
     *
     * @var string
     */
    protected $primaryKey = 'id';

    /**
     * 不使用Laravel的时间戳管理
     *
     * @var bool
     */
    public $timestamps = false;

    /**
     * 禁用软删除
     * wechat_message 表没有 deleted_at 字段
     *
     * @return string|null
     */
    public function getDeletedAtColumn(): ?string
    {
        return null;
    }

    /**
     * 可批量赋值的属性
     *
     * @var array
     */
    protected $fillable = [
        'msgid',
        'seq',
        'msgtime',
        'action',
        'roomid',
        'conversation_id',
        'conversation_type',
        'from_user',
        'to_list',
        'msgtype',
        'content',
        'raw_json',
        'resource_id',
    ];

    /**
     * 需要进行类型转换的属性
     *
     * @var array
     */
    protected $casts = [
        'id' => 'integer',
        'seq' => 'integer',
        'msgtime' => 'integer',
        'to_list' => 'array',
        'content' => 'array',
        'raw_json' => 'array',
    ];

    /**
     * 获取消息所属的会话
     * 定义多对一关系
     *
     * @return BelongsTo
     */
    public function conversation(): BelongsTo
    {
        return $this->belongsTo(Conversation::class, 'conversation_id', 'conversation_id');
    }

    /**
     * 获取关联的媒体资源
     * 定义多对一关系
     *
     * @return BelongsTo
     */
    public function resource(): BelongsTo
    {
        return $this->belongsTo(WeChatResource::class, 'resource_id', 'id');
    }

    /**
     * 根据会话ID获取聊天记录
     * 支持分页、搜索，关联查询 wechat_member 表获取发送者的昵称、备注名称、头像等信息
     *
     * @param string $conversationId 会话ID
     * @param int $page 页码
     * @param int $pageSize 每页数量
     * @param string|null $searchSender 搜索发送者（from_user、昵称或备注名称）
     * @param string|null $searchContent 搜索内容关键词
     * @return array 返回包含列表和分页信息的数组，每条消息包含 sender_nick_name, sender_remark_name, sender_avatar 等字段
     */
    public static function getMessagesByConversation(
        string $conversationId, 
        int $page = 1, 
        int $pageSize = 50,
        ?string $searchSender = null,
        ?string $searchContent = null
    ): array
    {
        $query = self::query()
                    ->withoutGlobalScopes() // 移除所有全局作用域（包括软删除）
                    ->select([
                        'wechat_message.*',
                        'wechat_member.nick_name as sender_nick_name',
                        'wechat_member.remark_name as sender_remark_name',
                        'wechat_member.type as sender_type',
                        'wechat_member.avatar as sender_avatar',
                        'wechat_member.gender as sender_gender',
                        'wechat_resource.oss_url as media_url'
                    ])
                    ->leftJoin('wechat_member', function($join) {
                        $join->on('wechat_message.from_user', '=', \DB::raw('wechat_member.account_name COLLATE utf8mb4_unicode_ci'));
                    })
                    ->leftJoin('wechat_resource', function($join) {
                        $join->on('wechat_message.resource_id', '=', 'wechat_resource.id')
                             ->where('wechat_resource.status', '=', 1);
                    })
                    ->where('wechat_message.conversation_id', $conversationId);

        // 搜索发送者（支持 from_user、nick_name 和 remark_name）
        if ($searchSender) {
            $query->where(function($q) use ($searchSender) {
                $q->where('wechat_message.from_user', 'like', '%' . $searchSender . '%')
                  ->orWhere('wechat_member.nick_name', 'like', '%' . $searchSender . '%')
                  ->orWhere('wechat_member.remark_name', 'like', '%' . $searchSender . '%');
            });
        }

        // 搜索内容（搜索 JSON 字段）
        if ($searchContent) {
            $query->where(function($q) use ($searchContent) {
                // 搜索文本消息内容
                $q->whereRaw("JSON_EXTRACT(content, '$.text.content') LIKE ?", ['%' . $searchContent . '%'])
                  // 或搜索文件名
                  ->orWhereRaw("JSON_EXTRACT(content, '$.file.filename') LIKE ?", ['%' . $searchContent . '%']);
            });
        }

        $query->orderByDesc('wechat_message.msgtime');

        // 统计总数（使用相同的查询条件）
        $countQuery = self::query()
                    ->withoutGlobalScopes()
                    ->where('conversation_id', $conversationId);

        if ($searchSender) {
            $countQuery->leftJoin('wechat_member', function($join) {
                $join->on('wechat_message.from_user', '=', \DB::raw('wechat_member.account_name COLLATE utf8mb4_unicode_ci'));
            })->where(function($q) use ($searchSender) {
                $q->where('wechat_message.from_user', 'like', '%' . $searchSender . '%')
                  ->orWhere('wechat_member.nick_name', 'like', '%' . $searchSender . '%')
                  ->orWhere('wechat_member.remark_name', 'like', '%' . $searchSender . '%');
            });
        }

        if ($searchContent) {
            $countQuery->where(function($q) use ($searchContent) {
                $q->whereRaw("JSON_EXTRACT(content, '$.text.content') LIKE ?", ['%' . $searchContent . '%'])
                  ->orWhereRaw("JSON_EXTRACT(content, '$.file.filename') LIKE ?", ['%' . $searchContent . '%']);
            });
        }

        $total = $countQuery->count();
                    
        $list = $query->offset(($page - 1) * $pageSize)
                     ->limit($pageSize)
                     ->get()
                     ->toArray();

        return [
            'list' => $list,
            'page' => $page,
            'pageSize' => $pageSize,
            'total' => $total,
        ];
    }

    /**
     * 根据消息ID获取消息详情
     *
     * @param string $msgid 消息ID
     * @return WeChatMessage|null
     */
    public static function getMessageDetail(string $msgid): ?WeChatMessage
    {
        return self::query()
            ->withoutGlobalScopes() // 移除所有全局作用域（包括软删除）
            ->where('msgid', $msgid)
            ->first();
    }

    /**
     * 获取消息统计信息
     *
     * @param string|null $conversationId 可选的会话ID
     * @return array
     */
    public static function getMessageStatistics(?string $conversationId = null): array
    {
        $query = self::query()
            ->withoutGlobalScopes(); // 移除所有全局作用域（包括软删除）

        if ($conversationId) {
            $query->where('conversation_id', $conversationId);
        }

        $stats = [
            'total_messages' => $query->count(),
        ];

        // 按消息类型统计
        $typeStats = (clone $query)->selectRaw('msgtype, COUNT(*) as count')
                                   ->groupBy('msgtype')
                                   ->get()
                                   ->pluck('count', 'msgtype')
                                   ->toArray();

        $stats['by_type'] = $typeStats;

        return $stats;
    }
}

File: collect_server/src/main/java/com/wechat/collector/model/WeChatMessage.java
package com.wechat.collector.model;

import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import lombok.Data;

import java.sql.Timestamp;
import java.util.Arrays;

/**
 * 企业微信消息实体类
 * 对应数据库表 wechat_message
 */
@Data
public class WeChatMessage {
    /**
     * 自增主键
     */
    private Long id;
    
    /**
     * 企业微信消息ID（唯一）
     */
    private String msgid;
    
    /**
     * 顺序号，拉取时排序用
     */
    private Long seq;
    
    /**
     * 消息时间戳（毫秒）
     */
    private Long msgtime;
    
    /**
     * 消息动作：send/recall
     */
    private String action;
    
    /**
     * 群聊ID（如果是群消息）
     */
    private String roomid;
    
    /**
     * 会话ID
     * 群聊：等于roomid
     * 单聊：两个userid按字典序排列，用下划线连接，如 "user1_user2"
     */
    private String conversationId;
    
    /**
     * 会话类型：single单聊 / group群聊
     */
    private String conversationType;
    
    /**
     * 发送者userid
     */
    private String fromUser;
    
    /**
     * 接收人列表（JSON格式）
     */
    private String toList;
    
    /**
     * 消息类型：text/image/voice/file/video/emotion/mixed等
     */
    private String msgtype;
    
    /**
     * 消息主要内容（JSON格式）
     */
    private String content;
    
    /**
     * 媒体文件OSS链接（image/video/voice/file类型消息）
     */
    private String mediaUrl;
    
    /**
     * 资源文件ID（关联wechat_resource表）
     */
    private Integer resourceId;
    
    /**
     * 原始微信返回JSON（可选）
     */
    private String rawJson;
    
    /**
     * 记录创建时间
     */
    private Timestamp createTime;
    
    /**
     * 生成会话ID和会话类型
     * 群聊：conversationId = roomid, conversationType = "group"
     * 单聊：conversationId = 两个userid按字典序排列（user1_user2）, conversationType = "single"
     */
    public void generateConversationInfo() {
        if (roomid != null && !roomid.isEmpty()) {
            // 群聊消息
            this.conversationId = roomid;
            this.conversationType = "group";
        } else {
            // 单聊消息：从toList中提取接收人，与发送人组合
            String toUser = extractFirstRecipient(this.toList);
            if (toUser != null) {
                this.conversationId = generateSingleConversationId(this.fromUser, toUser);
                this.conversationType = "single";
            }
        }
    }
    
    /**
     * 从toList JSON中提取第一个接收人
     * 
     * @param toListJson toList的JSON字符串，如 "[\"userid1\"]" 或 "[\"userid1\",\"userid2\"]"
     * @return 第一个接收人的userid，解析失败返回null
     */
    private String extractFirstRecipient(String toListJson) {
        if (toListJson == null || toListJson.isEmpty()) {
            return null;
        }
        
        try {
            ObjectMapper mapper = new ObjectMapper();
            JsonNode arrayNode = mapper.readTree(toListJson);
            
            if (arrayNode.isArray() && arrayNode.size() > 0) {
                return arrayNode.get(0).asText();
            }
        } catch (Exception e) {
            // 解析失败，返回null
        }
        
        return null;
    }
    
    /**
     * 生成单聊会话ID
     * 将两个userid按字典序排列，用下划线连接
     * 
     * @param user1 第一个用户ID
     * @param user2 第二个用户ID
     * @return 会话ID，如 "user1_user2"
     */
    private String generateSingleConversationId(String user1, String user2) {
        if (user1 == null || user2 == null) {
            return null;
        }
        
        // 按字典序排列，确保A与B聊天、B与A聊天的conversationId一致
        String[] users = {user1, user2};
        Arrays.sort(users);
        
        return users[0] + "_" + users[1];
    }
}

